File: Lab1/Python/task2_send_recv/send_recv.py

from mpi4py import MPI

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()

if size < 2:
    print("World size must be greater than 1")
    MPI.COMM_WORLD.Abort(1)

if rank == 0:
    message = "Hello Process 1!"
    comm.send(message, dest=1, tag=0)
    print("Process 0 sending message to process 1")
elif rank == 1:
    received_message = comm.recv(source=0, tag=0)
    print(f"Process 1 received message: {received_message}")
else:
    print(f"Process {rank} is not involved in this communication.")

File: Lab1/C/task2_send_recv/send_recv.c

#include <mpi.h>
#include <stdio.h>

int main(int argc, char** argv) {
    int world_rank, world_size, message, received_message;
    MPI_Status status;

    MPI_Init(NULL, NULL);
    MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
    MPI_Comm_size(MPI_COMM_WORLD, &world_size);

    if (world_size < 2) {
        fprintf(stderr, "World size must be greater than 1\n");
        MPI_Abort(MPI_COMM_WORLD, 1);
    }

    if (world_rank == 0) {
        message = 123;
        MPI_Send(&message, 1, MPI_INT, 1, 0, MPI_COMM_WORLD);
        printf("Process 0 sending message to process 1\n");
    } else if (world_rank == 1) {
        MPI_Recv(&received_message, 1, MPI_INT, 0, 0, MPI_COMM_WORLD, &status);
        printf("Process 1 received message: %d\n", received_message);
    } else {
        printf("Process %d is not involved in this communication.\n", world_rank);
    }

    MPI_Finalize();
    return 0;
}
